File: firefly/src/main/java/com/firefly/codec/http2/stream/SimpleFlowControlStrategy.java
package com.firefly.codec.http2.stream;

import com.firefly.codec.http2.frame.Frame;
import com.firefly.codec.http2.frame.WindowUpdateFrame;
import com.firefly.utils.concurrent.Callback;

public class SimpleFlowControlStrategy extends AbstractFlowControlStrategy {
	public SimpleFlowControlStrategy() {
		this(DEFAULT_WINDOW_SIZE);
	}

	public SimpleFlowControlStrategy(int initialStreamSendWindow) {
		super(initialStreamSendWindow);
	}

	@Override
	public void onDataConsumed(SessionSPI session, StreamSPI stream, int length) {
		if (length <= 0)
			return;

		// This is the simple algorithm for flow control.
		// This method is called when a whole flow controlled frame has been
		// consumed.
		// We send a WindowUpdate every time, even if the frame was very small.

		WindowUpdateFrame sessionFrame = new WindowUpdateFrame(0, length);
		session.updateRecvWindow(length);
		if (log.isDebugEnable())
			log.debug("Data consumed, increased session recv window by {} for {}", length, session);

		Frame[] streamFrame = Frame.EMPTY_ARRAY;
		if (stream != null) {
			if (stream.isClosed()) {
				if (log.isDebugEnable())
					log.debug("Data consumed, ignoring update stream recv window by {} for closed {}", length, stream);
			} else {
				streamFrame = new Frame[1];
				streamFrame[0] = new WindowUpdateFrame(stream.getId(), length);
				stream.updateRecvWindow(length);
				if (log.isDebugEnable())
					log.debug("Data consumed, increased stream recv window by {} for {}", length, stream);
			}
		}

		session.frames(stream, Callback.NOOP, sessionFrame, streamFrame);
	}
}

File: firefly-net/src/main/kotlin/com/fireflysource/net/http/common/v2/stream/SimpleFlowControlStrategy.kt
package com.fireflysource.net.http.common.v2.stream

import com.fireflysource.common.sys.SystemLogger
import com.fireflysource.net.http.common.HttpConfig
import com.fireflysource.net.http.common.v2.frame.WindowUpdateFrame

class SimpleFlowControlStrategy(
    initialStreamRecvWindow: Int = HttpConfig.DEFAULT_WINDOW_SIZE
) : AbstractFlowControlStrategy(initialStreamRecvWindow) {

    companion object {
        private val log = SystemLogger.create(SimpleFlowControlStrategy::class.java)
    }

    override fun onDataConsumed(http2Connection: Http2Connection, stream: Stream?, length: Int) {
        if (length <= 0) return

        // This is the simple algorithm for flow control.
        // This method is called when a whole flow controlled frame has been consumed.
        // We send a WindowUpdate every time, even if the frame was very small.
        val connection = http2Connection as AsyncHttp2Connection
        val sessionFrame = WindowUpdateFrame(0, length)
        connection.updateRecvWindow(length)
        log.debug { "Data consumed, increased session recv window by $length for $connection" }

        var streamFrame: WindowUpdateFrame? = null
        if (stream != null && stream is AsyncHttp2Stream) {
            if (stream.isRemotelyClosed()) {
                log.debug { "Data consumed, ignoring update stream recv window by $length for remotely closed $stream" }
            } else {
                streamFrame = WindowUpdateFrame(stream.id, length)
                stream.updateRecvWindow(length)
                log.debug { "Data consumed, increased stream recv window by $length for $stream" }
            }
        }
        if (streamFrame != null) {
            connection.sendControlFrame(stream, sessionFrame, streamFrame)
        } else {
            connection.sendControlFrame(stream, sessionFrame)
        }
    }

}